design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,146 @@
+services.
+
+### Token-based Auth
+
+Token-based auth allows users to use credentials like username and password to obtain a token which allows communicating
+with the system without using credentials.
+As Monolith/BFF performs authentication & authorization token-based auth allows using tokens for a secure communication
+between a user, Monolith/BFF, and services.
+
+[JWT](https://jwt.io/) could be used as a "secure" token implementation and the main benefits are:
+* Allows using adaptive structure to pass permissions between services
+* Could be encrypted (JWE) or just signed (JWS)
+* As it's RFC it has implementations for most of the modern programming languages, so there is no a hard
+dependency on some stack for the service implementation
+
+JWT has three parts separated by `.` (dots):
+* Header - contains a type of the token and the hashing algorithm
+* Payload - contains an information about the entity
+* Signature - encoded string for a token validation
+
+The signature encoded by HMAC SHA256 algorithm might look like this:
+```
+HMACSHA256(
+    base64UrlEncode(header) + "." +
+    base64UrlEncode(payload),
+    secret
+)
+```
+    
+## General Schema
+
+Now, current Magento authorization mechanism based on roles and assigned rules. Each role has a list of allowed resources
+and ER diagram looks like this:
+
+![Current RBAC ER](current-rbac-er.png)
+
+Where `resource_id` is represented as module name + resource name (`Magento_SalesArchive::remove`, `Magento_SalesArchive::remove`, etc.),
+and each resource has permission `allow` or `deny`. But current schema does not apply to WEB API because now, each WEB API
+entry point should specify the resource. Also, the permission column looks like obvious, if a resource is not specified for
+a rule when the resource is not allowed.
+
+It's better to expose all available API and use entry points instead of existing resources as rules. For example,
+`rest/V1/shipment/:id` and HTTP verb (like POST or GET), this approach is similar to [Amazon ARNs](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html)
+and allows making fine-grained access control.
+
+As Magento should provide [multiple BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends)
+it makes sense to extract authentication and authorization flow to a separate `Auth Service` to avoid credentials
+validation and duplication of roles & privileges data. And the general schema might look like this:
+
+![Authentication and Authorization General Schema](auth-general-schema.png)
+
+1. An unauthorized user sends a request with credentials to retrieve an access token, BFF/Monolith proxies this request to
+`Auth Service` which performs authorization & authentication and issues access token based on a user role.
+2. The user sends the request to get a needed resource with the access token, BFF/Monolith proxies this request to the appropriate
```

Comment:
Can BFF handle tokens? This would allow to save 1 network call and will not require clients know about tokens. To receive product information, for example, client would send request to BFF. BFF will request the token and send the request to Catalog service.

Reply:
Probably we need to discuss requests segregation to private and public again.